game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java

```diff
@@ -218,6 +238,7 @@ public boolean isSet() {
 
   @Override
   public void save(final String newValue) {
+    onSaveAction.accept(newValue);
```

Comment:
One way to test this is to reverse the dependency.  Instead of `ClientSetting` knowing what the save action is, allow it to be specified by an external client.  That's probably better for decoupling in the long run.  For example, I'm not sure if `ClientSetting` should know about `ErrorConsole`, much less contain behavior manipulating it.

Anyway, here's a rough cut at allowing multiple save listeners per setting (note that it leaves the current save actions in place and doesn't try to move them outside this type, but that would be the goal in the final realization):

```diff
diff --git a/game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java b/game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java
index 7c71bc164..03fc02039 100644
--- a/game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java
+++ b/game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java
@@ -3,6 +3,8 @@ package games.strategy.triplea.settings;
 import static com.google.common.base.Preconditions.checkNotNull;
 
 import java.io.File;
+import java.util.Collection;
+import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import java.util.prefs.BackingStoreException;
@@ -139,13 +141,13 @@ public enum ClientSetting implements GameSetting {
 
   private static final AtomicReference<Preferences> preferencesRef = new AtomicReference<>();
 
-  private final Consumer<String> onSaveAction;
+  private final Collection<Consumer<String>> saveListeners = new CopyOnWriteArrayList<>();
 
   public final String defaultValue;
 
   ClientSetting(final String defaultValue, final Consumer<String> onSaveAction) {
     this.defaultValue = defaultValue;
-    this.onSaveAction = onSaveAction;
+    saveListeners.add(onSaveAction);
   }
 
   ClientSetting(final boolean defaultVAlue, final Consumer<String> onSaveAction) {
@@ -238,7 +240,7 @@ public enum ClientSetting implements GameSetting {
 
   @Override
   public void save(final String newValue) {
-    onSaveAction.accept(newValue);
+    saveListeners.forEach(listener -> listener.accept(newValue));
     getPreferences().put(name(), newValue);
   }
 
@@ -264,4 +266,14 @@ public enum ClientSetting implements GameSetting {
   public void resetAndFlush() {
     saveAndFlush(defaultValue);
   }
+
+  @Override
+  public void addSaveListener(final Consumer<String> listener) {
+    saveListeners.add(listener);
+  }
+
+  @Override
+  public void removeSaveListener(final Consumer<String> listener) {
+    saveListeners.remove(listener);
+  }
 }
diff --git a/game-core/src/main/java/games/strategy/triplea/settings/GameSetting.java b/game-core/src/main/java/games/strategy/triplea/settings/GameSetting.java
index cac7322f2..163e1cf5c 100644
--- a/game-core/src/main/java/games/strategy/triplea/settings/GameSetting.java
+++ b/game-core/src/main/java/games/strategy/triplea/settings/GameSetting.java
@@ -1,5 +1,7 @@
 package games.strategy.triplea.settings;
 
+import java.util.function.Consumer;
+
 /**
  * Basic save/read API for a 'game setting'. Generally these will be values the user can modify over the
  * course of the game, either directly or indirectly. For example, default window size may saved here,
@@ -38,4 +40,8 @@ public interface GameSetting {
   }
 
   void resetAndFlush();
+
+  void addSaveListener(Consumer<String> listener);
+
+  void removeSaveListener(Consumer<String> listener);
 }
diff --git a/game-core/src/test/java/games/strategy/triplea/settings/ClientSettingTest.java b/game-core/src/test/java/games/strategy/triplea/settings/ClientSettingTest.java
new file mode 100644
index 000000000..3229bda90
--- /dev/null
+++ b/game-core/src/test/java/games/strategy/triplea/settings/ClientSettingTest.java
@@ -0,0 +1,38 @@
+package games.strategy.triplea.settings;
+
+import static org.mockito.Mockito.verify;
+
+import java.util.function.Consumer;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+
+import com.example.mockito.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+public final class ClientSettingTest extends AbstractClientSettingTestCase {
+  @Mock
+  private Consumer<String> saveListener;
+
+  @BeforeEach
+  public void addSaveListener() {
+    ClientSetting.TEST_SETTING.addSaveListener(saveListener);
+  }
+
+  @AfterEach
+  public void removeSaveListener() {
+    ClientSetting.TEST_SETTING.removeSaveListener(saveListener);
+  }
+
+  @Test
+  public void shouldInvokeSaveListenersWhenNewValueSaved() {
+    final String newValue = "newValue";
+
+    ClientSetting.TEST_SETTING.save(newValue);
+
+    verify(saveListener).accept(newValue);
+  }
+}
```

Reply:
Interesting. Looking at this closer, wouldn't every `Consumer<String>` be invoked every time?

Reply:
I am considering having for example ErrorConsole register with the right client setting. I'm not sure about this; i like the tight cohesion of having all related configurations together in one place. Ideally you could add an entire setting with a new enum entry. 

I'm going to attempt to restructure things so that ClientSetting configuration is offloaded to a builder that we could test.

Reply:
Not sure if I understand the question.  Every `Consumer<String>` for the specific `ClientSetting` whose value was changed would be invoked, yes.  That's no different than what's happening currently except that each `ClientSetting` can now have more than one listener.  I used a collection of listeners for each setting instead of one per setting because, if you move the listener implementations outside this type, you really don't know how many clients may register for notification of a setting change.

Reply:
My misunderstanding on that point, it is 1:1 by enum value and not all enums.

@ssoloff this is a good suggestion.

The one thing I have been wondering about is initialization references. `ErrorConsole` should work as we init it explicitly on startup. On the other hand other components that could have save actions, but are not yet initialized, could be missed.

Perhaps we can resolve this be extracting all such classes/initialization to a common spot and then initialize it all together in GameRunner.

Reply:
I gave the suggestion a try, the initialization ordering issue turned out to be a problem.

I'm also hesitant to let 'save actions' be registered from anywhere in the code base. This seems to be poor cohesion. It would probably be okay if all actions were localized to the same package, but otherwise if it really grew you would have a hard time tracking down every save action listener; would be hard to know what would happen if we did a reset to factory default.

Reply:
> I gave the suggestion a try, the initialization ordering issue turned out to be a problem.

Derp, need to click save. We probably should make it so settings just update as they are changed instead of needing the 'save'